src/Microsoft.Diagnostics.EventFlow.Inputs.Etw/EtwInput.cs

```diff
@@ -0,0 +1,163 @@
+            if (providersConfiguration == null)
+            {
+                healthReporter.ReportProblem($"{nameof(EtwInput)}: required configuration section 'providers' is missing");
+                return;
+            }
+            var providers = new List<EtwProviderConfiguration>();
+            try
+            {
+                providersConfiguration.Bind(providers);
+            }
+            catch
+            {
+                healthReporter.ReportProblem($"{nameof(EtwInput)}: configuration is invalid", EventFlowContextIdentifiers.Configuration);
+                return;
+            }
+
+            Initialize(providers, healthReporter);
+        }
+
+        public EtwInput(IReadOnlyCollection<EtwProviderConfiguration> providers, IHealthReporter healthReporter)
+        {
+            Requires.NotNull(providers, nameof(providers));
+            Requires.NotNull(healthReporter, nameof(healthReporter));
+
+            Initialize(providers, healthReporter);
+        }
+
+        public IEnumerable<EtwProviderConfiguration> Providers { get; private set; }
+        public Func<ITraceEventSession> SessionFactory { get; set; }
+
+        public void Activate()
+        {
+            if (this.isDisposed)
+            {
+                throw new ObjectDisposedException(nameof(EtwInput));
+            }
+
+            if (this.session != null)
+            {
+                throw new InvalidOperationException($"{nameof(EtwInput)} has already been activated");
+            }
+
+            if (this.Providers.Count() == 0)
+            {
+                healthReporter.ReportWarning($"{nameof(EtwInput)}: no providers configured", nameof(EtwInput));
+                return;
+            }
+
+            this.session = SessionFactory();
+
+            foreach (var providerConfiguration in this.Providers)
+            {
+                string validationError;
+                if (!providerConfiguration.Validate(out validationError))
+                {
+                    this.healthReporter.ReportWarning($"{nameof(EtwInput)}: one of the providers is configured improperly. {validationError}", EventFlowContextIdentifiers.Configuration);
+                    continue;
+                }
+                if (providerConfiguration.ProviderGuid != Guid.Empty)
+                {
+                    this.session.EnableProvider(providerConfiguration.ProviderGuid, providerConfiguration.Level, (ulong)providerConfiguration.Keywords);
+                }
+                else
+                {
+                    this.session.EnableProvider(providerConfiguration.ProviderName, providerConfiguration.Level, (ulong)providerConfiguration.Keywords);
+                }
+            }
+
+            // Also enable .NET Task Parallel Library hierarchical activity tracking
+            this.session.EnableProvider(TplActivities.TplEventSourceGuid, TraceEventLevel.Always, TplActivities.TaskFlowActivityIdsKeyword);
```

Comment:
Think about enabling this provider with this keyword as a flag that enables [hierarchical activity generation for .NET applications](https://blogs.msdn.microsoft.com/vancem/2015/09/14/exploring-eventsource-activity-correlation-and-causation-features/). This is low-cost thing, but maybe should be a "on by default, can turn off if necessary" thing. I will leave as-is for now.